src/mlpack/methods/kmeans/naive_kmeans_impl.hpp

```diff
@@ -38,28 +40,52 @@ double NaiveKMeans<MetricType, MatType>::Iterate(const arma::mat& centroids,
   counts.zeros(centroids.n_cols);
 
   // Find the closest centroid to each point and update the new centroids.
-  for (size_t i = 0; i < dataset.n_cols; i++)
+  // Computed in parallel over the complete dataset
+  #pragma omp parallel
   {
-    // Find the closest centroid to this point.
-    double minDistance = std::numeric_limits<double>::infinity();
-    size_t closestCluster = centroids.n_cols; // Invalid value.
+    // The current state of the K-means is private for each thread
+    arma::mat localCentroids(centroids.n_rows, centroids.n_cols,
+                             arma::fill::zeros);
+    arma::Col<size_t> localCounts(centroids.n_cols, arma::fill::zeros);
 
-    for (size_t j = 0; j < centroids.n_cols; j++)
+#ifdef _WIN32
+  // Tiny workaround: Visual Studio only implements OpenMP 2.0, which doesn't
+  // support unsigned loop variables. If we're building for Visual Studio, use
+  // the intmax_t type instead.
+  #pragma omp for
+    for (intmax_t i = 0; i < (intmax_t) dataset.n_cols; ++i)
+#else
+  #pragma omp for
+    for (size_t i = 0; i < dataset.n_cols; ++i)
+#endif
     {
-      const double distance = metric.Evaluate(dataset.col(i), centroids.col(j));
+      // Find the closest centroid to this point.
+      double minDistance = std::numeric_limits<double>::infinity();
+      size_t closestCluster = centroids.n_cols; // Invalid value.
 
-      if (distance < minDistance)
+      for (size_t j = 0; j < centroids.n_cols; j++)
       {
-        minDistance = distance;
-        closestCluster = j;
+        const double distance = metric.Evaluate(dataset.col(i),
+                                                centroids.unsafe_col(j));
+        if (distance < minDistance)
+        {
+          minDistance = distance;
+          closestCluster = j;
+        }
       }
-    }
 
-    Log::Assert(closestCluster != centroids.n_cols);
+      Log::Assert(closestCluster != centroids.n_cols);
 
-    // We now have the minimum distance centroid index.  Update that centroid.
-    newCentroids.col(closestCluster) += arma::vec(dataset.col(i));
-    counts(closestCluster)++;
+      // We now have the minimum distance centroid index.  Update that centroid.
+      localCentroids.unsafe_col(closestCluster) += dataset.col(i);
+      localCounts(closestCluster)++;
+    }
+    // Combine calculated state from each thread
+    #pragma omp critical
```

Comment:
Nice, thanks for doing the correct thing (profiling) and not what I did (guessing).

I'll take a closer look on the algorithm (this time, I solemnly swear to profile), and see if I can come up with any other potential optimizations.

However, 1.2x the Armadillo times is still a significant bump :)